athenz-identity-provider-service/src/main/java/com/yahoo/vespa/hosted/athenz/instanceproviderservice/AthenzSslKeyStoreConfigurator.java

```diff
@@ -2,73 +2,115 @@
     private final KeyProvider keyProvider;
     private final AthenzProviderServiceConfig.Zones zoneConfig;
-    private final AtomicBoolean alreadyConfigured = new AtomicBoolean();
+    private final Duration updatePeriod;
+    private final Path keystoreCachePath;
     private volatile KeyStore currentKeyStore;
 
     @Inject
     public AthenzSslKeyStoreConfigurator(KeyProvider keyProvider,
                                          AthenzProviderServiceConfig config,
-                                         Zone zone) {
+                                         Zone zone,
+                                         ConfigserverConfig configserverConfig) {
         AthenzProviderServiceConfig.Zones zoneConfig = getZoneConfig(config, zone);
-        this.certificateClient = new AthenzCertificateClient(config, zoneConfig);
+        Path keystoreCachePath = createKeystoreCachePath(configserverConfig);
+        AthenzCertificateClient certificateClient = new AthenzCertificateClient(config, zoneConfig);
+        Duration updatePeriod = Duration.ofDays(config.updatePeriodDays());
+        this.certificateClient = certificateClient;
         this.keyProvider = keyProvider;
         this.zoneConfig = zoneConfig;
-        this.currentKeyStore = downloadCertificate(keyProvider, certificateClient, zoneConfig);
+        this.currentKeyStore = initializeKeystore(keyProvider, certificateClient, zoneConfig, keystoreCachePath, updatePeriod);
+        this.updatePeriod = updatePeriod;
+        this.keystoreCachePath = keystoreCachePath;
+    }
+
+    private static KeyStore initializeKeystore(KeyProvider keyProvider,
+                                               AthenzCertificateClient certificateClient,
+                                               AthenzProviderServiceConfig.Zones zoneConfig,
+                                               Path keystoreCachePath,
+                                               Duration updatePeriod) {
+        return tryReadKeystoreFile(keystoreCachePath.toFile(), updatePeriod)
+                .orElseGet(() -> downloadCertificate(keyProvider, certificateClient, zoneConfig, keystoreCachePath));
+    }
+
+    private static Optional<KeyStore> tryReadKeystoreFile(File certificateFile, Duration updatePeriod) {
+        try {
+            if (!certificateFile.exists()) return Optional.empty();
+            KeyStore keyStore = KeyStore.getInstance("JKS");
+            try (InputStream in = new BufferedInputStream(new FileInputStream(certificateFile))) {
+                keyStore.load(in, new char[0]);
+            }
+            Instant minimumExpiration = Instant.now().plus(updatePeriod).plus(EXPIRATION_MARGIN);
+            boolean isExpired = getKeyStoreExpiry(keyStore).isBefore(minimumExpiration);
+            if (isExpired) return Optional.empty();
+            return Optional.of(keyStore);
+        } catch (IOException | GeneralSecurityException e) {
+            throw new RuntimeException(e);
```

Comment:
What should happen if reading the keystore fails? Currently this will fail starting the config server. Intentional?

Reply:
Currently yes. I am open to change this semantic.